<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>libPoMo.vcf &mdash; PoMo 1.1.0 documentation</title>
    
    <link rel="stylesheet" href="_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="PoMo 1.1.0 documentation" href="index.html" />
    <link rel="next" title="libPoMo.cf" href="cf.html" />
    <link rel="prev" title="libPoMo.fasta" href="fasta.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="cf.html" title="libPoMo.cf"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="fasta.html" title="libPoMo.fasta"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">PoMo 1.1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="module-libPoMo.vcf"></span><div class="section" id="libpomo-vcf">
<h1>libPoMo.vcf<a class="headerlink" href="#libpomo-vcf" title="Permalink to this headline">¶</a></h1>
<p>This module provides functions to read, write and access vcf files.</p>
<div class="section" id="objects">
<h2>Objects<a class="headerlink" href="#objects" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt>Classes:</dt>
<dd><ul class="first last simple">
<li><a class="reference internal" href="#libPoMo.vcf.NucBase" title="libPoMo.vcf.NucBase"><code class="xref py py-class docutils literal"><span class="pre">NucBase</span></code></a>, store a nucleotide base</li>
<li><a class="reference internal" href="#libPoMo.vcf.VCFStream" title="libPoMo.vcf.VCFStream"><code class="xref py py-class docutils literal"><span class="pre">VCFStream</span></code></a>, a variant call format (VCF) stream object</li>
<li><a class="reference internal" href="#libPoMo.vcf.VCFSeq" title="libPoMo.vcf.VCFSeq"><code class="xref py py-class docutils literal"><span class="pre">VCFSeq</span></code></a>, a VCF file sequence object</li>
</ul>
</dd>
<dt>Exception Classes:</dt>
<dd><ul class="first last simple">
<li><a class="reference internal" href="#libPoMo.vcf.NotAVariantCallFormatFileError" title="libPoMo.vcf.NotAVariantCallFormatFileError"><code class="xref py py-class docutils literal"><span class="pre">NotAVariantCallFormatFileError</span></code></a></li>
<li><a class="reference internal" href="#libPoMo.vcf.NotANucBaseError" title="libPoMo.vcf.NotANucBaseError"><code class="xref py py-class docutils literal"><span class="pre">NotANucBaseError</span></code></a></li>
</ul>
</dd>
<dt>Functions:</dt>
<dd><ul class="first last simple">
<li><a class="reference internal" href="#libPoMo.vcf.update_base" title="libPoMo.vcf.update_base"><code class="xref py py-func docutils literal"><span class="pre">update_base()</span></code></a>, read a line into a base</li>
<li><a class="reference internal" href="#libPoMo.vcf.get_nuc_base_from_line" title="libPoMo.vcf.get_nuc_base_from_line"><code class="xref py py-func docutils literal"><span class="pre">get_nuc_base_from_line()</span></code></a>, create a new <cite>NucBase</cite> from a line</li>
<li><a class="reference internal" href="#libPoMo.vcf.check_fixed_field_header" title="libPoMo.vcf.check_fixed_field_header"><code class="xref py py-func docutils literal"><span class="pre">check_fixed_field_header()</span></code></a>, check a VCF fixed field header
string</li>
<li><a class="reference internal" href="#libPoMo.vcf.get_indiv_from_field_header" title="libPoMo.vcf.get_indiv_from_field_header"><code class="xref py py-func docutils literal"><span class="pre">get_indiv_from_field_header()</span></code></a>, extract list of individuals
from header</li>
<li><a class="reference internal" href="#libPoMo.vcf.init_seq" title="libPoMo.vcf.init_seq"><code class="xref py py-func docutils literal"><span class="pre">init_seq()</span></code></a>, open VCF file and initialize <cite>VCFStream</cite></li>
<li><a class="reference internal" href="#libPoMo.vcf.open_seq" title="libPoMo.vcf.open_seq"><code class="xref py py-func docutils literal"><span class="pre">open_seq()</span></code></a>, open VCF file and save it to a <cite>VCFSeq</cite></li>
<li><a class="reference internal" href="#libPoMo.vcf.get_header_line_string" title="libPoMo.vcf.get_header_line_string"><code class="xref py py-func docutils literal"><span class="pre">get_header_line_string()</span></code></a>, print vcf header line</li>
</ul>
</dd>
</dl>
<hr class="docutils" />
<dl class="exception">
<dt id="libPoMo.vcf.NotANucBaseError">
<em class="property">exception </em><code class="descclassname">libPoMo.vcf.</code><code class="descname">NotANucBaseError</code><a class="reference internal" href="_modules/libPoMo/vcf.html#NotANucBaseError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.NotANucBaseError" title="Permalink to this definition">¶</a></dt>
<dd><p>Exception raised if given nucleotide base is not valid.</p>
</dd></dl>

<dl class="exception">
<dt id="libPoMo.vcf.NotAVariantCallFormatFileError">
<em class="property">exception </em><code class="descclassname">libPoMo.vcf.</code><code class="descname">NotAVariantCallFormatFileError</code><a class="reference internal" href="_modules/libPoMo/vcf.html#NotAVariantCallFormatFileError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.NotAVariantCallFormatFileError" title="Permalink to this definition">¶</a></dt>
<dd><p>Exception raised if given VCF file is not valid.</p>
</dd></dl>

<dl class="class">
<dt id="libPoMo.vcf.NucBase">
<em class="property">class </em><code class="descclassname">libPoMo.vcf.</code><code class="descname">NucBase</code><a class="reference internal" href="_modules/libPoMo/vcf.html#NucBase"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.NucBase" title="Permalink to this definition">¶</a></dt>
<dd><p>Stores a nucleotide base.</p>
<p>FIXME: Bases are split by &#8216;/&#8217;.  They should also be split by &#8216;|&#8217;.</p>
<p>A class that stores a single nucleotide base and related
information retrieved from a VCF file.  Please see
<a class="reference external" href="http://www.1000genomes.org/">http://www.1000genomes.org/</a> for a detailed description of the vcf
format.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first last simple">
<li><strong>chrom</strong> (<em>str</em>) &#8211; Chromosome name.</li>
<li><strong>pos</strong> (<em>int</em>) &#8211; 1-based position on the chromosome.</li>
<li><strong>id</strong> (<em>str</em>) &#8211; ID.</li>
<li><strong>ref</strong> (<em>str</em>) &#8211; Reference base.</li>
<li><strong>alt</strong> (<em>str</em>) &#8211; Alternative base(s).</li>
<li><strong>qual</strong> (<em>str</em>) &#8211; Quality.</li>
<li><strong>filter</strong> (<em>str</em>) &#8211; Filter.</li>
<li><strong>info</strong> (<em>str</em>) &#8211; Additional information.</li>
<li><strong>format</strong> (<em>str</em>) &#8211; String with format specification.</li>
<li><strong>speciesData</strong> (<em>[str]</em>) &#8211; List with strings of the species data
(e.g. 0/1:...).</li>
<li><strong>ploidy</strong> (<em>int</em>) &#8211; Ploidy (number of sets of chromosomes) of the
sequenced individuals. Can be set with
<a class="reference internal" href="#libPoMo.vcf.NucBase.set_ploidy" title="libPoMo.vcf.NucBase.set_ploidy"><code class="xref py py-func docutils literal"><span class="pre">set_ploidy()</span></code></a>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="libPoMo.vcf.NucBase.get_alt_base_list">
<code class="descname">get_alt_base_list</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#NucBase.get_alt_base_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.NucBase.get_alt_base_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Return alternative bases as a list.</p>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.NucBase.get_base_ind">
<code class="descname">get_base_ind</code><span class="sig-paren">(</span><em>iI</em>, <em>iC</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#NucBase.get_base_ind"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.NucBase.get_base_ind" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the base of a specific individual.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>indiv</strong> (<em>int</em>) &#8211; 0-based index of individual.</li>
<li><strong>chrom</strong> (<em>int</em>) &#8211; 0-based index of chromosome (for n-ploid
individuals).</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">character with nucleotide base.</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.NucBase.get_info">
<code class="descname">get_info</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#NucBase.get_info"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.NucBase.get_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Return nucleotide base information string.</p>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.NucBase.get_ref_base">
<code class="descname">get_ref_base</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#NucBase.get_ref_base"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.NucBase.get_ref_base" title="Permalink to this definition">¶</a></dt>
<dd><p>Return reference base.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">char</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.NucBase.get_speciesData">
<code class="descname">get_speciesData</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#NucBase.get_speciesData"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.NucBase.get_speciesData" title="Permalink to this definition">¶</a></dt>
<dd><p>Return species data as a list.</p>
<ul>
<li><p class="first">data[0][0] = data of first species/individual on chromatide A</p>
</li>
<li><dl class="first docutils">
<dt>data[0][1] = only set for non-haploids; data of first</dt>
<dd><p class="first last">species/individual on chromatide B</p>
</dd>
</dl>
</li>
</ul>
<p>Sets data[i][j] to None if the base of individual <em>i</em> on
chromosome <em>j</em> could not be read (e.g. it is not valid).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">matrix of integers</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.NucBase.print_info">
<code class="descname">print_info</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#NucBase.print_info"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.NucBase.print_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Print nucleotide base information.</p>
<p>Print the stored single nucleotide base and related
information from the VCF file.</p>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.NucBase.purge">
<code class="descname">purge</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#NucBase.purge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.NucBase.purge" title="Permalink to this definition">¶</a></dt>
<dd><p>Purge the data associated with this <a class="reference internal" href="#libPoMo.vcf.NucBase" title="libPoMo.vcf.NucBase"><code class="xref py py-class docutils literal"><span class="pre">NucBase</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.NucBase.set_ploidy">
<code class="descname">set_ploidy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#NucBase.set_ploidy"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.NucBase.set_ploidy" title="Permalink to this definition">¶</a></dt>
<dd><p>Set self.ploidy.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="libPoMo.vcf.VCFSeq">
<em class="property">class </em><code class="descclassname">libPoMo.vcf.</code><code class="descname">VCFSeq</code><a class="reference internal" href="_modules/libPoMo/vcf.html#VCFSeq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.VCFSeq" title="Permalink to this definition">¶</a></dt>
<dd><p>Store data retrieved from a VCF file.</p>
<p>Initialized with <a class="reference internal" href="#libPoMo.vcf.open_seq" title="libPoMo.vcf.open_seq"><code class="xref py py-func docutils literal"><span class="pre">open_seq()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> (<em>str</em>) &#8211; Sequence name.</li>
<li><strong>header</strong> (<em>str</em>) &#8211; Sequence header.</li>
<li><strong>speciesL</strong> (<em>[str]</em>) &#8211; List with species / individuals.</li>
<li><strong>nSpecies</strong> (<em>int</em>) &#8211; Number of species / individuals.</li>
<li><strong>baseL</strong> (<em>[NucBase]</em>) &#8211; List with stored <a class="reference internal" href="#libPoMo.vcf.NucBase" title="libPoMo.vcf.NucBase"><code class="xref py py-class docutils literal"><span class="pre">NucBase</span></code></a> objects.</li>
<li><strong>nBases</strong> (<em>int</em>) &#8211; Number of <a class="reference internal" href="#libPoMo.vcf.NucBase" title="libPoMo.vcf.NucBase"><code class="xref py py-class docutils literal"><span class="pre">NucBase</span></code></a> objects stored.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="libPoMo.vcf.VCFSeq.append_nuc_base">
<code class="descname">append_nuc_base</code><span class="sig-paren">(</span><em>base</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#VCFSeq.append_nuc_base"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.VCFSeq.append_nuc_base" title="Permalink to this definition">¶</a></dt>
<dd><p>Append <em>base</em>, a given <a class="reference internal" href="#libPoMo.vcf.NucBase" title="libPoMo.vcf.NucBase"><code class="xref py py-class docutils literal"><span class="pre">NucBase</span></code></a>, to the VCFSeq object.</p>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.VCFSeq.get_header_line_string">
<code class="descname">get_header_line_string</code><span class="sig-paren">(</span><em>indiv</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#VCFSeq.get_header_line_string"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.VCFSeq.get_header_line_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a standard VCF File header string with individuals <em>indiv</em>.</p>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.VCFSeq.get_nuc_base">
<code class="descname">get_nuc_base</code><span class="sig-paren">(</span><em>chrom</em>, <em>pos</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#VCFSeq.get_nuc_base"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.VCFSeq.get_nuc_base" title="Permalink to this definition">¶</a></dt>
<dd><p>Return base at position <em>pos</em> of chromosome <em>chrom</em>.</p>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.VCFSeq.has_base">
<code class="descname">has_base</code><span class="sig-paren">(</span><em>chrom</em>, <em>pos</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#VCFSeq.has_base"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.VCFSeq.has_base" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True (False) if base is (not) found.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>chrom</strong> (<em>str</em>) &#8211; Chromosome name.</li>
<li><strong>pos</strong> (<em>int</em>) &#8211; 1-based position on <em>chrom</em>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.VCFSeq.print_header_line">
<code class="descname">print_header_line</code><span class="sig-paren">(</span><em>indiv</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#VCFSeq.print_header_line"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.VCFSeq.print_header_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Print a standard VCF File header with individuals <em>indiv</em>.</p>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.VCFSeq.print_info">
<code class="descname">print_info</code><span class="sig-paren">(</span><em>maxB=50</em>, <em>printHeader=False</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#VCFSeq.print_info"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.VCFSeq.print_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Print VCF sequence information.</p>
<p>Print vcf header, the total number of nucleotides and a
maximum of <em>maxB</em> bases (defaults to 50). Only prints header
if <em>printHeader</em> = True is given.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="libPoMo.vcf.VCFStream">
<em class="property">class </em><code class="descclassname">libPoMo.vcf.</code><code class="descname">VCFStream</code><span class="sig-paren">(</span><em>seqName</em>, <em>vcfFileObject</em>, <em>speciesList</em>, <em>firstBase</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#VCFStream"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.VCFStream" title="Permalink to this definition">¶</a></dt>
<dd><p>Store base data from a VCF file line per line.</p>
<p>It can be initialized with <a class="reference internal" href="#libPoMo.vcf.init_seq" title="libPoMo.vcf.init_seq"><code class="xref py py-func docutils literal"><span class="pre">init_seq()</span></code></a>.  This class stores a
single base retrieved from a VCF file and the file itself.  It is
used to parse through a VCF file line by line processing the bases
without having to read the whole file at one.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>seqName</strong> (<em>str</em>) &#8211; Name of the stream.</li>
<li><strong>vcfFileObject</strong> (<em>fo</em>) &#8211; File object associated with the stream.</li>
<li><strong>speciesList</strong> (<em>[str]</em>) &#8211; List with species / individuals.</li>
<li><strong>firstBase</strong> (<a class="reference internal" href="#libPoMo.vcf.NucBase" title="libPoMo.vcf.NucBase"><em>NucBase</em></a>) &#8211; First <a class="reference internal" href="#libPoMo.vcf.NucBase" title="libPoMo.vcf.NucBase"><code class="xref py py-class docutils literal"><span class="pre">NucBase</span></code></a> to be saved.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Variables:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> (<em>str</em>) &#8211; Name of the stream.</li>
<li><strong>fo</strong> (<em>fo</em>) &#8211; Stored VCF file object.</li>
<li><strong>speciesL</strong> (<em>[str]</em>) &#8211; List with species / individuals.</li>
<li><strong>nSpecies</strong> (<em>int</em>) &#8211; Number of species / individuals.</li>
<li><strong>base</strong> (<em>NusBase</em>) &#8211; Stored <a class="reference internal" href="#libPoMo.vcf.NucBase" title="libPoMo.vcf.NucBase"><code class="xref py py-class docutils literal"><span class="pre">NucBase</span></code></a>.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="libPoMo.vcf.VCFStream.close">
<code class="descname">close</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#VCFStream.close"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.VCFStream.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Closes the linked file.</p>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.VCFStream.print_info">
<code class="descname">print_info</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#VCFStream.print_info"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.VCFStream.print_info" title="Permalink to this definition">¶</a></dt>
<dd><p>Prints VCFStream information.</p>
</dd></dl>

<dl class="method">
<dt id="libPoMo.vcf.VCFStream.read_next_base">
<code class="descname">read_next_base</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#VCFStream.read_next_base"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.VCFStream.read_next_base" title="Permalink to this definition">¶</a></dt>
<dd><p>Read the next base.</p>
<p>Return position of next base.</p>
<p>Raise a <em>ValueError</em> if no next base is found.</p>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="libPoMo.vcf.check_fixed_field_header">
<code class="descclassname">libPoMo.vcf.</code><code class="descname">check_fixed_field_header</code><span class="sig-paren">(</span><em>ln</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#check_fixed_field_header"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.check_fixed_field_header" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the given line <em>ln</em> is the header of the fixed fields.</p>
<p>Sample header line:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1">#CHROM     POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  SpeciesL</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="libPoMo.vcf.get_header_line_string">
<code class="descclassname">libPoMo.vcf.</code><code class="descname">get_header_line_string</code><span class="sig-paren">(</span><em>indiv</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#get_header_line_string"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.get_header_line_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a standard VCF File header string with individuals <em>indiv</em>.</p>
</dd></dl>

<dl class="function">
<dt id="libPoMo.vcf.get_indiv_from_field_header">
<code class="descclassname">libPoMo.vcf.</code><code class="descname">get_indiv_from_field_header</code><span class="sig-paren">(</span><em>ln</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#get_indiv_from_field_header"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.get_indiv_from_field_header" title="Permalink to this definition">¶</a></dt>
<dd><p>Return species from a fixed field header line <em>ln</em>.</p>
<p>Sample header line:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="c1">#CHROM     POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  SpeciesL</span>
</pre></div>
</div>
</dd></dl>

<dl class="function">
<dt id="libPoMo.vcf.get_nuc_base_from_line">
<code class="descclassname">libPoMo.vcf.</code><code class="descname">get_nuc_base_from_line</code><span class="sig-paren">(</span><em>ln</em>, <em>info=False</em>, <em>ploidy=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#get_nuc_base_from_line"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.get_nuc_base_from_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve base data from a VCF file line <em>ln</em>.</p>
<p>Split a given VCF file line and returns a NucBase object. If
<em>info</em> is set to False, only #CHROM, POS, REF, ALT and speciesData will
be read.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>info</strong> (<em>Bool</em>) &#8211; Determines if info is retrieved from <em>ln</em>.</li>
<li><strong>ploidy</strong> (<em>int</em>) &#8211; If ploidy is known and given, it is set.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="libPoMo.vcf.init_seq">
<code class="descclassname">libPoMo.vcf.</code><code class="descname">init_seq</code><span class="sig-paren">(</span><em>VCFFileName</em>, <em>maxskip=100</em>, <em>name=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#init_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.init_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Open a (gzipped) VCF4.1 file.</p>
<p>Try to open the given VCF file, checks if it is in VCF format.
Initialize a <a class="reference internal" href="#libPoMo.vcf.VCFStream" title="libPoMo.vcf.VCFStream"><code class="xref py py-class docutils literal"><span class="pre">VCFStream</span></code></a> object that contains the first
base.</p>
<p>Please close the associated file object with
<a class="reference internal" href="#libPoMo.vcf.VCFStream.close" title="libPoMo.vcf.VCFStream.close"><code class="xref py py-func docutils literal"><span class="pre">VCFStream.close()</span></code></a> when you don&#8217;t need it anymore.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>VCFFileName</strong> (<em>str</em>) &#8211; Name of the VCF file.</li>
<li><strong>maxskip</strong> (<em>int</em>) &#8211; Only look <em>maxskip</em> lines for the start of the
bases (defaults to 80).</li>
<li><strong>name</strong> (<em>str</em>) &#8211; Set the name of the sequence to <em>name</em>, otherwise
set it to the filename.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="libPoMo.vcf.open_seq">
<code class="descclassname">libPoMo.vcf.</code><code class="descname">open_seq</code><span class="sig-paren">(</span><em>VCFFileName</em>, <em>maxskip=100</em>, <em>name=None</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#open_seq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.open_seq" title="Permalink to this definition">¶</a></dt>
<dd><p>Open a VCF4.1 file.</p>
<p>Try to open the given VCF file, checks if it is in VCF format and
reads the bases(s).  It returns an <a class="reference internal" href="#libPoMo.vcf.VCFSeq" title="libPoMo.vcf.VCFSeq"><code class="xref py py-class docutils literal"><span class="pre">VCFSeq</span></code></a> object that
contains all the information.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>VCFFileName</strong> (<em>str</em>) &#8211; Name of the VCF file.</li>
<li><strong>maxskip</strong> (<em>int</em>) &#8211; Only look <em>maxskip</em> lines for the start of the
bases (defaults to 80).</li>
<li><strong>name</strong> (<em>str</em>) &#8211; Set the name of the sequence to <em>name</em>, otherwise
set it to the filename.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="libPoMo.vcf.update_base">
<code class="descclassname">libPoMo.vcf.</code><code class="descname">update_base</code><span class="sig-paren">(</span><em>ln</em>, <em>base</em>, <em>info=True</em><span class="sig-paren">)</span><a class="reference internal" href="_modules/libPoMo/vcf.html#update_base"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#libPoMo.vcf.update_base" title="Permalink to this definition">¶</a></dt>
<dd><p>Read line <em>ln</em> into base <em>base</em>.</p>
<p>Split a given VCF file line and returns a <a class="reference internal" href="#libPoMo.vcf.NucBase" title="libPoMo.vcf.NucBase"><code class="xref py py-class docutils literal"><span class="pre">NucBase</span></code></a>
object. If <em>info</em> is set to False, only #CHROM, REF, ALT and
speciesData will be read.</p>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">libPoMo.vcf</a><ul>
<li><a class="reference internal" href="#objects">Objects</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="fasta.html"
                        title="previous chapter">libPoMo.fasta</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="cf.html"
                        title="next chapter">libPoMo.cf</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/vcf.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="cf.html" title="libPoMo.cf"
             >next</a> |</li>
        <li class="right" >
          <a href="fasta.html" title="libPoMo.fasta"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">PoMo 1.1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2015, Dominik Schrempf.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.5.
    </div>
  </body>
</html>