
The following (cleaned) files provided information about the study cohort
dataset for \VAR{ initial_nb_samples } samples and \VAR{ initial_nb_markers }
markers (including \VAR{ nb_special_markers } markers located on sexual or
mitochondrial chromosomes):

\begin{itemize}
\BLOCK{ for data_file in data_files }
  \item \path{\VAR{ data_file }}
\BLOCK{ endfor }
\end{itemize}

IMPUTE2's pre-phasing approach can work with phased haplotypes from SHAPEIT, a
highly accurate phasing algorithm that can handle mixtures of unrelated
samples, duos or trios. The usage of SHAPEIT is highly recommended to infer
haplotypes underlying the study genotypes. The phased haplotypes are then
passed to IMPUTE2 for imputation. Although pre-phasing allows for very fast
imputation, it leads to a small loss in accuracy since the estimation
uncertainty in the study haplotypes is ignored. SHAPEIT version
\VAR{ shapeit_version }~\cite{Delaneau13_23269371} and IMPUTE2 version
\VAR{ impute2_version }~\cite{Howie09_19543373,Howie11_22384356,Howie12_22820512}
were used for this analysis. Binary pedfiles were processed using Plink version
\VAR{ plink_version }~\cite{Purcell07_17701901}.\\

To speed up the pre-phasing and imputation steps, the dataset was split by
chromosome. The following quality steps were then performed on each chromosome:

\VAR{ steps_data }

\BLOCK{ if no_marker_left|length > 0 }
After exclusion, chromosome\BLOCK{ if no_marker_left|length > 1}s\BLOCK{ endif }
 \VAR{ no_marker_left[:-1]|join(", ") }\BLOCK{ if no_marker_left|length > 1 }
 and \BLOCK{ endif }\VAR{ no_marker_left[-1] } did not have enough markers and
\BLOCK{ if no_marker_left|length > 1 }were \BLOCK{ else }was \BLOCK{ endif }
skipped for downstream analysis.\\
\BLOCK{ endif }

\textbf{In total, \VAR{ nb_phasing_markers } markers were used for phasing
using SHAPEIT.} IMPUTE2 was then used to impute markers genome-wide using its
reference file\VAR{ filtering_rules }.\BLOCK{ if nb_samples_no_gender != "0" } \textbf{A
total of \VAR{ nb_samples_no_gender }
sample\VAR{ "" if nb_samples_no_gender == "1" else "s" } did not have gender
information, and were skip for chromosome X
imputation.}\BLOCK{ endif }\BLOCK{ if no_imputed_sites|length > 0 }\\

Chromosome\BLOCK{ if no_imputed_sites|length > 1}s\BLOCK{ endif }
 \VAR{ no_imputed_sites[:-1]|join(", ") }\BLOCK{ if no_imputed_sites|length > 1 }
 and \BLOCK{ endif }\VAR{ no_imputed_sites[-1] } did not have sufficient
markers to perform imputation. Hence,
\BLOCK{ if no_imputed_sites|length > 1 }they were \BLOCK{ else }it was \BLOCK{ endif }
skipped for downstream analysis.\\
\BLOCK{ endif }

